File: src/main/java/com/kika/smllybot/modules/user/Local/ui/ProfileUI.java
package com.kika.smllybot.modules.user.Local.ui;

import com.kika.smllybot.modules.user.Local.ProfileContext;
import com.kika.smllybot.utils.TimeUtil;
import net.dv8tion.jda.api.components.container.Container;
import net.dv8tion.jda.api.components.container.ContainerChildComponent;
import net.dv8tion.jda.api.components.section.Section;
import net.dv8tion.jda.api.components.separator.Separator;
import net.dv8tion.jda.api.components.textdisplay.TextDisplay;
import net.dv8tion.jda.api.components.thumbnail.Thumbnail;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ProfileUI {

    public static Container buildProfile(ProfileContext ctx) {
        List<ContainerChildComponent> components = new ArrayList<>();

        String avatarUrl = ctx.target().getEffectiveAvatarUrl();
        String name = ctx.target().getEffectiveName();
        String timestamp = TimeUtil.getTimestamp(ctx.profile().getCreatedAt());
        String timestampRelative = TimeUtil.getTimestampRelative(ctx.profile().getCreatedAt());

        ContainerChildComponent header = Section.of(
                Thumbnail.fromUrl(avatarUrl),
                TextDisplay.of("# \\👤 Это %s".formatted(name)),
                TextDisplay.of("### О себе:"),
                TextDisplay.of("%s".formatted(ctx.profile().getAboutMe()))
        );
        ContainerChildComponent separator = Separator.createDivider(Separator.Spacing.SMALL);
        ContainerChildComponent main = TextDisplay.of("## Основная информация:");
        ContainerChildComponent owner = TextDisplay.of("\\👑 Создатель дискорд-сервера");
        ContainerChildComponent citizen = TextDisplay.of("\\🪪 Гражданин этого сервера");
        ContainerChildComponent reputation = TextDisplay.of("Репутация: \\✨ %d | \\➕ %d"
                .formatted(ctx.bank().getStar(), ctx.user().getReaction()));
        ContainerChildComponent firstEntry = TextDisplay.of("Впервые появился здесь %s (%s)"
                .formatted(timestamp, timestampRelative));
        ContainerChildComponent activity = TextDisplay.of("**Актив** (д | н | м | весь): %d | %d | %d | %d"
                .formatted(ctx.statistic().getDay(), ctx.statistic().getWeek(), ctx.statistic().getMonth(), ctx.statistic().getTotal()));
        ContainerChildComponent rewardsHeader = TextDisplay.of("## \\🏆 Награды");
        ContainerChildComponent rewards = TextDisplay.of("...");
        ContainerChildComponent footer = TextDisplay
                .of("-# Это - **локальный** профиль. Часть информации выводится исключительно из этой гильдии");

        components.add(header);
        components.add(separator);
        components.add(main);
        if (ctx.member().isOwner()) components.add(owner);
        if (Objects.equals(ctx.user().getCitizenship(), ctx.guildId())) components.add(citizen);
        components.add(reputation);
        components.add(firstEntry);
        components.add(activity);
        components.add(separator);
        components.add(rewardsHeader);
        components.add(rewards);
        components.add(separator);
        components.add(footer);

        return Container.of(components);
    }
}

File: src/main/java/com/kika/smllybot/modules/user/Local/ui/ProfileUI.kt
package com.kika.smllybot.modules.user.Local.ui

import com.kika.smllybot.modules.user.Local.ProfileContext
import com.kika.smllybot.utils.NumUtil
import com.kika.smllybot.utils.TimeUtil
import net.dv8tion.jda.api.components.actionrow.ActionRow
import net.dv8tion.jda.api.components.buttons.Button
import net.dv8tion.jda.api.components.buttons.ButtonStyle
import net.dv8tion.jda.api.components.container.Container
import net.dv8tion.jda.api.components.container.ContainerChildComponent
import net.dv8tion.jda.api.components.section.Section
import net.dv8tion.jda.api.components.separator.Separator
import net.dv8tion.jda.api.components.textdisplay.TextDisplay
import net.dv8tion.jda.api.components.thumbnail.Thumbnail

class ProfileUI {

    companion object {
        @JvmStatic
        fun buildProfile(ctx: ProfileContext): Container {
            val components: MutableList<ContainerChildComponent?> = ArrayList()

            val avatarUrl = ctx.target.getEffectiveAvatarUrl()
            val name = ctx.target.getEffectiveName()
            val timestamp = TimeUtil.getTimestamp(ctx.profile.createdAt)
            val timestampRelative = TimeUtil.getTimestampRelative(ctx.profile.createdAt)
            val reaction = NumUtil.german(ctx.user.reaction)
            val star = NumUtil.german(ctx.bank.star)
            val day = NumUtil.german(ctx.statistic.day)
            val week = NumUtil.german(ctx.statistic.week)
            val month = NumUtil.german(ctx.statistic.month)
            val total = NumUtil.german(ctx.statistic.total)
            val aboutMe = ctx.profile.aboutMe?.replace("@", "\\@") ?: "Пользователь не указал описание"

            val header: ContainerChildComponent = Section.of(
                Thumbnail.fromUrl(avatarUrl),
                TextDisplay.of("# \\👤 Это ${name.replace("@", "\\@")}"),
                TextDisplay.of("### О себе:"),
                TextDisplay.of(aboutMe)
            )
            val separator: ContainerChildComponent = Separator.createDivider(Separator.Spacing.SMALL)
            val main: ContainerChildComponent = TextDisplay.of("## Основная информация:")
            val owner: ContainerChildComponent = TextDisplay.of("\\👑 Создатель дискорд-сервера")
            val citizen: ContainerChildComponent = TextDisplay.of("\\🪪 Гражданин этого сервера")
            val reputation: ContainerChildComponent = TextDisplay.of(
                "Репутация: \\✨ $star | \\➕ $reaction"
            )
            val firstEntry: ContainerChildComponent = TextDisplay.of(
                "Впервые появился здесь $timestamp ($timestampRelative)"
            )
            val activity: ContainerChildComponent = TextDisplay.of(
                "**Актив** (д | н | м | весь): $day | $week | $month | $total"
            )
            val rewardsHeader: ContainerChildComponent = TextDisplay.of("## \\🏆 Награды")
            val rewards: ContainerChildComponent = TextDisplay.of("...")
            val buttonsHeader = TextDisplay.of("### Настройки профиля")
            val button = ActionRow.of(Button.of(ButtonStyle.PRIMARY, "profile:modal:${ctx.target.idLong}", "ℹ️ Редактировать о себе"))
            val footer: ContainerChildComponent = TextDisplay
                .of("-# Это - **локальный** профиль. Часть информации выводится исключительно из этой гильдии")

            components.add(header)
            components.add(separator)
            components.add(main)
            if (ctx.member.isOwner) components.add(owner)
            if (ctx.user.citizenship == ctx.guildId) components.add(citizen)
            components.add(reputation)
            components.add(firstEntry)
            components.add(activity)
//            components.add(rewardsHeader)
//            components.add(rewards)
            if (ctx.target.idLong == ctx.viewer.idLong) {
                components.add(separator)
                components.add(buttonsHeader)
                components.add(button)
            }
            components.add(separator)
            components.add(footer)

            return Container.of(components)
        }
    }
}